New API: Add DelegationRole and Delegations classes

MVrachev opened the pull request:
Addresses #1139

**Description of the changes being introduced by the pull request**:

  
  In the top-level metadata classes, there are complex attributes such as
  "meta" in Targets and Snapshot, "key" and "roles" in Root etc.
  We want to represent those complex attributes with a class to allow
  easier verification and support for metadata with unrecognized fields.
  For more context read ADR 0004 and ADR 0008 in the docs/adr folder.
  
  DelegationRole shares a couple of fields with the Role class and that's
  why it inherits it.
  I decided to use a separate Delegations class because I thought it will
  make it easier to read, verify and add additional helper functions.
  Also, I tried to make sure that I test each level of the delegations
  representation for support of storing unrecognized fields.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Add DelegatedRole and Delegations

In the top level metadata classes, there are complex attributes such as
"meta" in Targets and Snapshot, "key" and "roles" in Root etc.
We want to represent those complex attributes with a class to allow
easier verification and support for metadata with unrecognized fields.
For more context read ADR 0004 and ADR 0008 in the docs/adr folder.

DelegatedRole shares a couple of fields with the Role class and that's
why it inherits it.
I decided to use a separate Delegations class because I thought it will
make it easier to read, verify and add additional helper functions.
Also, I tried to make sure that I test each level of the delegations
representation for support of storing unrecognized fields.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Add Delegation/DelegatednRole specific tests

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+    @classmethod
+    def from_dict(cls, delegations_dict: Mapping[str, Any]) -> "Delegations":
+        """Creates Delegations object from its dict representation."""
+        keys = delegations_dict.pop("keys")
+        for keyid, key_dict in keys.items():
+            keys[keyid] = Key.from_dict(key_dict.copy())
```
modifying a dictionary while iterating through it is traditionally a red flag... maybe worth building a new dict?

MVrachev replied:
I am returning a new dictionary by calling `key_dict.copy()`.
Should I add a new variable to make it clearer?

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+            keys[keyid] = Key.from_dict(key_dict.copy())
+        roles = delegations_dict.pop("roles")
+        roles_res = []
+        for role_dict in roles:
+            new_role = DelegationRole.from_dict(role_dict.copy())
+            roles_res.append(new_role)
```
I don't think you need a copy() here?

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+            "roles": roles,
+            **self.unrecognized_fields,
+        }
+        return res_dict
```
nit: could just return dict without variable

jku commented on tuf/api/metadata.py:
```diff
@@ -808,17 +907,18 @@ def from_dict(cls, targets_dict: Mapping[str, Any]) -> "Targets":
         common_args = cls._common_fields_from_dict(targets_dict)
         targets = targets_dict.pop("targets")
-        delegations = targets_dict.pop("delegations")
+        delegations_dict = targets_dict.pop("delegations")
```
if delegations is optional then this is clearly a bug.

This is properly fixed in #1367

jku commented on tuf/api/metadata.py:
```diff
@@ -808,17 +907,18 @@ def from_dict(cls, targets_dict: Mapping[str, Any]) -> "Targets":
             {
                 "targets": self.targets,
-                "delegations": self.delegations,
+                "delegations": self.delegations.to_dict(),
```
bug when delegations is None: This is properly fixed in #1367

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns the dict representation of self."""
+        keys = {}
+        for keyid, key in self.keys.items():
+            keys[keyid] = key.to_dict()
+        roles = []
+        for role_obj in self.roles:
+            roles.append(role_obj.to_dict())
```
could consider dict/list comprehension again but up to you really

MVrachev replied:
Will try them.

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+                "Only one of the attributes 'paths' and"
+                "'path_hash_prefixes' can be set!"
+            )
+        if paths:
+            self.paths = paths
+        elif path_hash_prefixes:
+            self.path_hash_prefixes = path_hash_prefixes
```
if-else leads to undefined object attributes. Some variant of this is not tested.

MVrachev replied:
I don't think there is a mistake here.
From Joshua's comment [here](https://github.com/theupdateframework/specification/issues/156#issuecomment-823447318), I understand we can have one of them or either of them. The checks I made are:
- we can't have both of them (first if)
- we can have only `paths` (second if)
- we can have only `path_hash_prefixes` (the elif)
- we can have an object without any of them.

What I miss is a specific test for creating `DelegationRole` only with one of `paths` and `path_hash_prefixes`.

jku replied:
the instance attributes are left unset so I think `to_dict() `will fail in some of those cases (at least when you only have path_hash_prefixes). I don't think there should be any if-else here (except for the validation that raises)

MVrachev replied:
Just tested it and you are right. We don't need the if-else here.

MVrachev commented on tests/test_api.py:
```diff
@@ -441,6 +503,15 @@ def test_support_for_unrecognized_fields(self):
                     dict1["signed"]["keys"][keyid]["d"] = "c"
                 for role_str in dict1["signed"]["roles"].keys():
                     dict1["signed"]["roles"][role_str]["e"] = "g"
+            elif metadata == "targets" and dict1["signed"].get("delegations"):
+                for keyid in dict1["signed"]["delegations"]["keys"].keys():
+                    dict1["signed"]["delegations"]["keys"][keyid]["d"] = "c"
+                new_roles = []
+                for role in dict1["signed"]["delegations"]["roles"]:
+                    role["e"] = "g"
+                    new_roles.append(role)
+                dict1["signed"]["delegations"]["roles"] = new_roles
+                dict1["signed"]["delegations"]["foo"] = "bar"
```
I know this is ugly, but until @jku proposes his test where he automatically validates support for unrecognized fields on each level of the dictionary, we would have to stick to this.

jku replied:
the reason I haven't done it is that while there seemingly was a consensus that "unrecognised keys should be allowed everywhere in the file format" ... it's not really true. There's a lot of places where that would be weird and unsafe (what if someone inserts random data into "keys" or "roles") and we should not allow that. The real rule is probably "unreconised keys should be allowed at every _object_ root level": so you should be allowed to insert a key into Signed, Key or Role ... but not into any dictionary where we cannot know the recognised key names before hand (like "keys") or where the allowed keys are strictly defined (like "meta").

Poisoning everything except a hard-coded list of dictionaries (instead of only poisoning specific dicts) may still be the better testing solution but it's not trivial.

jku replied:
there's an implementation in #1382 now

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+    def __init__(
+        self,
+        name: str,
+        keyids: list,
+        threshold: int,
+        terminating: bool,
+        paths: Optional[list] = None,
+        path_hash_prefixes: Optional[list] = None,
```
three list type hints without list-item type hints

jku commented on tuf/api/metadata.py:
```diff
@@ -728,6 +728,135 @@ def update(
+    def __init__(
+        self,
+        keys: Mapping[str, Key],
+        roles: list,
```
missing list-item hint

jku reviewed (changes requested):
I'll take another look after #1367 fixes are done and this is rebased: currently it looks buggy

Commit:
Remove forgotten debug line from a test

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

Commit:
Don't use Mapping for mutable dicts

Stop using Mapping where we actually mean Dict:
Mapping means "we only need a read-only dict" and most of the time
this is not really the case.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Rebased on top of the latest develop branch and addressed all comments made by @jku.

jku commented on tuf/api/metadata.py:
```diff
@@ -740,6 +740,129 @@ def update(
+        keyids = role_dict.pop("keyids")
+        threshold = role_dict.pop("threshold")
+        terminating = role_dict.pop("terminating")
+        paths = paths = role_dict.pop("paths", None)
```
```suggestion
        paths = role_dict.pop("paths", None)
```

jku reviewed (commented):
This looks good to me, thanks.

I left one typo comment and have a naming question but it's basically :heavy_check_mark: for me.

Naming: Would DelegatedRole be better than DelegationRole -- maybe it conveys a little bit more information (in that this clearly is the role that is delegated, not the one that is delegating)? I don't feel strongly about it though, just want to make sure we try to find the best names...

MVrachev commented:
> This looks good to me, thanks.
> 
> I left one typo comment and have a naming question but it's basically ✔️ for me.
> 
> Naming: Would DelegatedRole be better than DelegationRole -- maybe it conveys a little bit more information (in that this clearly is the role that is delegated, not the one that is delegating)? I don't feel strongly about it though, just want to make sure we try to find the best names...

Yes, I agree `DelegatedRole` sounds better. I was never in love with the `DelegationRole` naming but wasn't sure what other options there are.

jku reviewed (approved):
Thanks!

MVrachev commented:
I rebased to fix one type-annotation I have forgotten.

MVrachev commented:
Rebased on top of the latest master branch.
